### PR TITLE
Authenication

### PR DESCRIPTION
**Pull request by Akola-Mbey-Denis**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -18,6 +18,12 @@ pyjwt = "^1.7.1"
 flask-restx = "^0.2.0"
 python-dotenv = "^0.13.0"
 pymysql = "^0.9.3"
+wheel = "^0.34.2"
+firebase-admin = "^4.3.0"
+grpcio = "^1.29.0"
+google-cloud-storage = "^1.28.1"
+setuptools = "^47.1.1"
+gcloud = "^0.18.3"
```
Comment (kakiang) [inline pyproject.toml:26]: You code makes use of all of these packages?

File: sme_financing/main/apis/fund_disbursement_api.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    @api.doc("Create a new fund disbursement")
+    @api.expect(_fund_disbursement, validate=True)
+    @api.response(201, "Funding disbursement successfully created")
+    def post(self):
+        """
+        Create a new funding criteria
+        """
+        data = request.json
+        return save_disbursement(data=data)
```
Comment (kakiang) [inline sme_financing/main/apis/fund_disbursement_api.py:35]: `FundDisbursement` is always linked to a `FundingProject`; it doesn't look like you have a way of getting a `funding projec`t here. 'cause given the `fund disbursement dto` you've defined, the `data` object won't contain a `funding  project` identifier. the endpoint for creating a fund disbursement should look like this `funding_projects/{funding_project_number}/fund_disbursements`. That way you get the project number from url.

File: sme_financing/main/service/fund_disbursement_service.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    db.session.commit()
+
+
+def commit_changes(data):
+    db.session.add(data)
+    update()
+
+
+def save_disbursement(data):
+    funding_project = get_funding_project_by_id(id=data["funding_project_id"])
```
Comment (kakiang) [inline sme_financing/main/service/fund_disbursement_service.py:21]: Not sure the `data` object will have `funding_project_id` key

File: sme_financing/main/api_v1.py [hunk 4]
```diff
@@ -3,11 +3,15 @@
 
 from .apis.client_api import api as client_ns
 from .apis.document_api import api as document_ns
+from .apis.fund_disbursement_api import api as disburse_ns
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/api_v1.py:6]: okay

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply pyproject.toml:26]: no please

File: sme_financing/main/api_v1.py [hunk 5]
```diff
@@ -26,6 +30,10 @@
 rest_api.add_namespace(sme_ns, path="/smes")
 rest_api.add_namespace(document_ns, path="/documents")
 rest_api.add_namespace(funding_ns, path="/funding")
-rest_api.add_namespace(funding_criteria_ns)
+rest_api.add_namespace(funding_criteria_ns, path="/funding-criteria")
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/api_v1.py:33]: Kindly update them

File: sme_financing/main/api_v1.py [hunk 6]
```diff
@@ -26,6 +30,10 @@
...
+rest_api.add_namespace(login_ns, path="/login")
+rest_api.add_namespace(signup_ns, path="/register")
+rest_api.add_namespace(reset_ns, path="/reset")
+rest_api.add_namespace(disburse_ns, path="/disburse")
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/api_v1.py:37]: Kindly update it

File: sme_financing/main/apis/dto.py [hunk 7]
```diff
@@ -199,6 +199,22 @@ class FundingDTO:
...
+        {
+            "title": fields.String(required=True, description="Funding Criteria title"),
+            "description": fields.String(
+                required=True, description=" Description of funding criteria"
+            ),
+            "investor_id": fields.Integer(required=True),
+        },
+    )
+
+
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/apis/dto.py:217]: okay

File: sme_financing/main/apis/fund_disbursement_api.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    save_disbursement,
+    update_fund_disbursement,
+)
+from .dto import FundDisbursementDTO
+
+api = FundDisbursementDTO.fund_disbursement_api
+_fund_disbursement = FundDisbursementDTO.fund_disburse
+
+
+@api.route("/fund_disburse")
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/apis/fund_disbursement_api.py:18]: I don't understand it

File: sme_financing/main/models/funding_detail.py [hunk 9]
```diff
@@ -12,7 +12,7 @@ class FundingDetail(db.Model):
     title = db.Column(db.String(255), nullable=False)
     description = db.Column(db.String(255), nullable=True)
 
-    documents = db.relationship("Document", backref="funding_detail")
+    # documents = db.relationship("Document", backref="funding_detail")
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/models/funding_detail.py:15]: I had a problem working with the document methods,I wanted to even let you,I saw you had already implemented the fund detail.
Kindly refactor it

File: sme_financing/main/apis/funding_criteria_api.py [hunk 10]
```diff
@@ -9,17 +9,17 @@
...
 )
-from .dto import FundingDTO
+from .dto import FundingCriteriaDTO
 
-api = FundingDTO.funding_api
-_funding_criteria = FundingDTO.funding_criteria
-_funding_criteria_display = FundingDTO.funding_criteria_display
+api = FundingCriteriaDTO.funding_criteria_api
+_funding_criteria = FundingCriteriaDTO.funding_criteria
+# _funding_criteria_display = FundingDTO.funding_criteria_display
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/apis/funding_criteria_api.py:16]: I didn't understand the name spacing at that time. It was not easy for me to understand

File: sme_financing/main/apis/funding_criteria_api.py [hunk 11]
```diff
@@ -9,17 +9,17 @@
...
 
 
 @api.route("/funding_criteria")
 class fundingCriteriaResource(Resource):
     @api.doc("List all funding criteria")
-    @api.marshal_list_with(_funding_criteria_display, envelope="data")
+    @api.marshal_list_with(_funding_criteria, envelope="data")
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/apis/funding_criteria_api.py:22]: same

File: sme_financing/main/service/funding_criteria_service.py [hunk 12]
```diff
@@ -2,43 +2,49 @@
...
-    else:
-        funding_criteria.investor = investor
+        fund_criteria = FundingCriteria(
+            title=data["title"],
+            description=data["description"],
+            investor_id=data["investor_id"],
+        )
+
+        fund_criteria.investor = investor
+
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/service/funding_criteria_service.py:34]: I did not initially see the email field in the investor model. So I perceived it was a typo

File: sme_financing/main/service/funding_criteria_service.py [hunk 13]
```diff
@@ -47,33 +53,32 @@ def update_funding_criteria(data, funding_criteria):
...
+            update()
+            response_object = {
+                "status": "success",
+                "message": "Successfully updated.",
+            }
+            return response_object, 201
+        except SQLAlchemyError as err:
+            db.session.rollback()
+            response_object = {"status": "error", "message": str(err)}
+            return response_object, 400
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/service/funding_criteria_service.py:81]: same

File: sme_financing/main/apis/dto.py [hunk 14]
```diff
@@ -235,6 +251,62 @@ class FundingProjectDTO:
...
+            "bank_account_details_from": fields.String(
+                required=True, description="bank  account details from"
+            ),
+            "password": fields.String(required=True, description="new password"),
+            "cheque_details": fields.String(
+                required=True, description="cheque details"
+            ),
+        },
+    )
+
```
[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/apis/dto.py:308]: okay

[the remark this answers is not in the document]
Comment (Akola-Mbey-Denis) [reply sme_financing/main/apis/dto.py:308]: okay

File: sme_financing/main/apis/fund_disbursement_api.py (hunk 2 above)
Comment (Akola-Mbey-Denis) [reply sme_financing/main/apis/fund_disbursement_api.py:35]: sure

File: sme_financing/main/service/fund_disbursement_service.py (hunk 3 above)
Comment (Akola-Mbey-Denis) [reply sme_financing/main/service/fund_disbursement_service.py:21]: Yeah

**Comment by Akola-Mbey-Denis** [comment]
Kindly fix all the issues.

**Pull request merged**